hy/compiler.py

```diff
@@ -2163,7 +2163,12 @@ def nan(): return asty.BinOp(
 
     @builds(HySymbol)
     def compile_symbol(self, symbol):
-        if "." in symbol:
+        if symbol == "...":
+            if not PY3:
+                raise HyTypeError(symbol, 'ellipsis syntax is only supported '
```

Comment:
Although the `...` syntax is only supported on Python3, the `Ellipsis` object exists in 2.x (https://docs.python.org/2.4/lib/bltin-ellipsis-object.html). Couldn't we do:

```
if PY3:
    return asty.Ellipsis(symbol)
else:
    symbol = "Ellipsis"
```

Reply:
The output should probably be `__builtins__.Ellipsis`, since `Ellipsis` isn't a reserved word.

Reply:
Good call

Reply:
Unfortunately, it doesn't work. When you try to use Ellipsis inside a slice in python 2 that's been generated this way, python throws `ValueError: invalid integer value: None` when trying to load the ast.

The reason, I think, is because that patch generates code to load `Ellipsis`: 

```
slice=Index(value=Tuple(elts=[Num(n=1L), Name(id='Ellipsis', ctx=Load()), Num(n=6L)], ctx=Load()))
```

However, `Ellipsis` **is** a thing an Python 2's ast. This is what I get from ast.parse -> ast.dump on Python 2:

```
slice=ExtSlice(dims=[Index(value=Num(n=1)), Ellipsis(), Index(value=Num(n=6))])
```

Reply:
So why not use `ast.Ellipsis`, just like in Python 2?

Reply:
Now, the reason I gated this to Python2 is that there's some deeper issue with constructing Ellipsis on Python2:

```
⤷ TypeError: Can't add <hy.compiler.Result object at 0x7fc4c1834e30> with non-compiler result <_ast.Ellipsis object at 0x7fc4c17caa50>
```

I'll dig into what's going on.

Reply:
Possibly Python 2 will only allow `ast.Ellipsis` inside a slice. In that case, you'll probably need to edit `compile_cut_expression` to peek at its arguments and make an `ast.Ellipsis` itself, whereas `compile_symbol`  will need to create `__builtins__.Ellipsis` or raise an error.

Reply:
Yeah, so the problem is in python 2, `ast.Ellipsis` is a subclass of ast.slice, not ast.expr, and `Result.__add__` then doesn't know what do with it.

But the problem with that is it means its only valid as a dim inside `ExtSlice`, which we don't generate.

I think this is only going to work on Python 3. Which personally I'm fine with. If it where up to me, I'd be seriously looking at dropping Python 2 support 😉

Reply:
So we'll need to produce an ExtSlice under Python 2, right?

> If it where up to me, I'd be seriously looking at dropping Python 2 support

[Me too, but it isn't.](https://github.com/hylang/hy/issues/1217)

Reply:
@Kodiologist yeah, I think so. With @rkday proposed patch, the following snippet only partially works:

```clojure
(defn test-slice []
  (defclass EllipsisTest []
    (defn --getitem-- [self [start ellipsis end]]
      (assert (= ellipsis Ellipsis))
      (list (range start end))))

  (assert (= (get (EllipsisTest) (, 1 ... 6))
             [1 2 3 4 5])))
```

When manually copy and pasted into hy's REPL, it works as expected. When imported, Python decides it doesn't like the generated AST.

Which really gaslighted me while trying to dig into it, heh.

> Me too, but it isn't.

Maybe I should pipe up on there, but I think its worth reconsidering now. Do one more release for both Python 2 and Python 3, then drop Python 2 for the next one.

Python 3 only has got some serious momentum now (http://www.python3statement.org/) and maybe it makes sense to follow the scientific computing's timeline. I mean, IPython is Python 3 only now, and IPython Notebook is going to be as well by June this year.

But alas, that's neither here nor there.